sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala

```diff
@@ -668,13 +668,37 @@ case class StreamingSymmetricHashJoinExec(
       private val iteratorNotEmpty: Boolean = super.hasNext
 
       override def completion(): Unit = {
-        val isLeftSemiWithMatch =
-          joinType == LeftSemi && joinSide == LeftSide && iteratorNotEmpty
-        // Add to state store only if both removal predicates do not match,
-        // and the row is not matched for left side of left semi join.
-        val shouldAddToState =
-          !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow) &&
-          !isLeftSemiWithMatch
```

Comment:
I imagine that you considered constructing these predicates (e.g. `stateKeyWatermarkPredicateFunc` and `stateValueWatermarkPredicateFunc`) based off the watermark for late events, not the watermark for eviction. I spent some time working this out myself, and I know the subtle reason why this won't work, but I wanted to verify that you also considered this.

Reply:
It's neither watermark for late event and watermark for eviction - stream-stream join has its own "state watermark", which is going to be an "output watermark" for stream-stream join. The predicate is relying on state watermark.

Reply:
Right, but I had initially wondered why we can't construct the state watermark predicates using the watermark for late events.

For example, with the join predicate `L > R + 10`, we construct the state watermark to be `L <= watermark_for_eviction(R) + 10`, and I had initially thought that the fix for this correctness bug wasn't `L <= watermark_for_late_events(R) + 10`.

Reply:
Yes, just simply saying, the reason we break down watermark for late event and eviction applies to stream-stream join, except the fact we'll need to buffer event longer than watermark based on join condition, hence necessity of state watermark (and it's related to eviction).